File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/TaggedObjectInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1Encodable
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DLTaggedObject

@Serializable
data class TaggedObjectInfo(
  val explicit: Boolean,
  val tagNo: Int,
  val content: EncodableInfo
) : EncodableInfo {
  companion object {
    fun getTaggedObject(
      explicit: Boolean,
      tagNo: Int,
      content: ASN1Encodable
    ): DLTaggedObject = DLTaggedObject(explicit, tagNo, content)
  }

  override fun toPrimitive(): ASN1Primitive = getTaggedObject(explicit, tagNo, content.toPrimitive()).toASN1Primitive()

}

File: net/JetBrains.Serialization/src/TaggedObjectInfo.cs
using System;
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class TaggedObjectInfo : IEncodableInfo
{
  [JsonProperty("Explicit")] private bool _explicit;
  [JsonProperty("TagNo")] private int _tagNo;
  [JsonProperty("Content")] private IEncodableInfo _content;

  [JsonConstructor]
  public TaggedObjectInfo(bool explicitness, int tagNo, IEncodableInfo content)
  {
    _explicit = explicitness;
    _tagNo = tagNo;
    _content = content;
  }

  public static DerTaggedObject GetTaggedObject(bool explicitness, int tagNo, Asn1Encodable content)
  {
    return new DerTaggedObject(explicitness, tagNo, content);
  }

  public Asn1Encodable ToPrimitive()
  {
    return GetTaggedObject(_explicit, _tagNo, _content.ToPrimitive()).ToAsn1Object();
  }
}